lib/Dialect/MSFT/MSFTPasses.cpp

```diff
@@ -247,6 +248,176 @@ std::unique_ptr<Pass> createLowerToHWPass() {
 } // namespace msft
 } // namespace circt
 
+namespace {
+struct PartitionPass : public PartitionBase<PartitionPass> {
+  void runOnOperation() override;
+
+  void partition(MSFTModuleOp mod);
+  void partition(DesignPartitionOp part, SmallVectorImpl<Operation *> &users);
+
+private:
+  hw::SymbolCache topLevelSyms;
+};
+} // anonymous namespace
+
+/// Fill a symbol cache with all the top level symbols.
+static void populateSymbolCache(mlir::ModuleOp mod, hw::SymbolCache &cache) {
+  for (Operation &op : mod.getBody()->getOperations()) {
+    StringAttr symName = SymbolTable::getSymbolName(&op);
+    if (!symName)
+      continue;
+    // Add the symbol to the cache.
+    cache.addDefinition(symName, &op);
+  }
+  cache.freeze();
+}
+
+void PartitionPass::runOnOperation() {
+  ModuleOp outerMod = getOperation();
+  ::populateSymbolCache(outerMod, topLevelSyms);
+
+  // TODO: sort modules by number of instantiations so we only have to run one
+  // "bubbling" pass.
+  outerMod.walk([this](MSFTModuleOp mod) { partition(mod); });
+}
+
+void PartitionPass::partition(MSFTModuleOp mod) {
+  DenseMap<SymbolRefAttr, SmallVector<Operation *, 4>> partMembers;
+  mod.walk([&partMembers](Operation *op) {
+    if (auto part = op->getAttrOfType<SymbolRefAttr>("targetDesignPartition"))
+      partMembers[part].push_back(op);
+  });
+
+  SmallVector<DesignPartitionOp, 2> parts(mod.getOps<DesignPartitionOp>());
+  for (auto part : parts) {
+    SymbolRefAttr partSym = SymbolRefAttr::get(SymbolTable::getSymbolName(mod),
+                                               {SymbolRefAttr::get(part)});
+
+    auto usersIter = partMembers.find(partSym);
+    if (usersIter != partMembers.end()) {
+      this->partition(part, usersIter->second);
+      partMembers.erase(usersIter);
+    }
+    part.erase();
+  }
+
+  // TODO: For operations which target partitions not in the same module, bubble
+  // them up.
+}
+
+void PartitionPass::partition(DesignPartitionOp partOp,
+                              SmallVectorImpl<Operation *> &toMove) {
+
+  auto *ctxt = partOp.getContext();
+  auto loc = partOp.getLoc();
```

Comment:
If everything will use the design partition's location, you can use an ImplicitLocOpBuilder and not specify the loc each time.

Reply:
Yeah, but things are built at three different places which means three builders anyway. (Or three setInsertionPoint calls on the same builder.)

Reply:
Oh, I meant that you just wouldn't need to pass "loc" to the calls to create, but I guess it wouldn't make much difference in this pass.